core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,387 @@
+import org.infinispan.util.concurrent.locks.PendingLockManager;
+import org.infinispan.util.concurrent.locks.PendingLockPromise;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.infinispan.commons.util.Util.toStr;
+import static org.infinispan.factories.KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR;
+
+/**
+ * The default implementation for {@link PendingLockManager}.
+ * <p>
+ * In transactional caches, a transaction would wait for transaction originated in a older topology id. It can happen
+ * when topology changes and a backup owner becomes the primary owner.
+ *
+ * @author Pedro Ruivo
+ * @since 8.0
+ */
+public class DefaultPendingLockManager implements PendingLockManager {
+
+   private static final Log log = LogFactory.getLog(DefaultPendingLockManager.class);
+   private static final boolean trace = log.isTraceEnabled();
+   private static final int NO_PENDING_CHECK = -2;
+   private final Map<GlobalTransaction, PendingLockPromiseImpl> pendingLockPromiseMap;
+   private TransactionTable transactionTable;
+   private TimeService timeService;
+   private ScheduledExecutorService timeoutExecutor;
+
+   public DefaultPendingLockManager() {
+      pendingLockPromiseMap = new ConcurrentHashMap<>();
+   }
+
+   @Inject
+   public void inject(TransactionTable transactionTable, TimeService timeService,
+                      @ComponentName(TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService timeoutExecutor) {
+      this.transactionTable = transactionTable;
+      this.timeService = timeService;
+      this.timeoutExecutor = timeoutExecutor;
+   }
+
+   @Override
+   public PendingLockPromise checkPendingForKey(TxInvocationContext<?> ctx, Object key, long time, TimeUnit unit) {
+      final GlobalTransaction globalTransaction = ctx.getGlobalTransaction();
+      PendingLockPromiseImpl existing = pendingLockPromiseMap.get(globalTransaction);
+      if (existing != null) {
+         return existing;
+      }
+      final int txTopologyId = getTopologyId(ctx);
+      if (txTopologyId != NO_PENDING_CHECK) {
+         if (trace) {
+            log.tracef("Checking for pending locks and then locking key %s", toStr(key));
+         }
+         return createAndStore(getTransactionWithLockedKey(txTopologyId, key, globalTransaction),
+                               globalTransaction, time, unit);
+      }
+      return PendingLockPromise.NO_OP;
```

Comment:
it is not a problem because I don't check the remaining timeout when it is NO_OP. and NO_OP means it hasn't wait anything :)